Add support for other export arguments than flags

adamuso opened the pull request:
Hello, 
I was missing support for other export arguments from GDScript (which can be seen in this [example](https://docs.godotengine.org/pl/stable/getting_started/scripting/gdscript/gdscript_exports.html)) so I have implemented additional decorator `export_args` which generates an export with custom arguments for a properties and getter/setter. Additionally I fixed `export_flags` because it was not working for getters and setters. Code that was written is mainly a copy of `export_flags` implementation with little tweaks. 

This was mainly done for my own needs, but I tried to make changes that fits your code style and I do not mind sharing :). Any feedback would be greatly appreciated.

ksjogo commented on parse_node/parse_class_declaration.ts:
```diff
@@ -140,6 +154,29 @@ class Foo {
+  ts: `
+@autoload
+class Foo {
+  @export_args(PackedScene)
```
Could we not make @exports take optional arguments, so we don't introduce another keyword?
Additionally wouldn't we be able to infer the type here instead of needing to manually provide it?
Afaict the codebase tried to avoid manual typing.

johnfn replied:
👍 to this comment. @exports taking optional arguments would be the ideal here!

adamuso replied:
Do we care about backwards compatibility? Because if we add arguments (even optional) to `exports` we cannot use it as we used to. 

```ts
// exports has no arguments, currently this code is valid
export class X extends Node2D {
  @exports
  test: int = 0;
}

// when exports definition is changed to accept arguments we need to add () to the end 
// of exports to make TypeScript not complain, this is not backward compatible
export class X extends Node2D {
  @exports()
  test: int = 0;
}
```

Also about infering the type, we can do that for array and maps but what about other cases like `DIR`, `MULTILINE`, `RGB`, etc. ?

adamuso replied:
> Do we care about backwards compatibility? Because if we add arguments (even optional) to exports we cannot use it as we used to.

Actually, I just tested it and if you make exports definition `declare function exports(...args: any[]): (target: Node, name: string) => void` then TypeScript does not care if it is really invoked or not. That means current code will also work. Sorry for misleading :)

ksjogo replied:
I actually didn't use DIR, Multiline etc. yet. That would indeed be more complicated.
So if no arg is provided, we infer and otherwise we use what's provided?

adamuso replied:
I've decided to infer and also add provided arguments after the inferred ones. See newest commits :)

ksjogo commented on parse_node/parse_class_declaration.ts:
```diff
@@ -39,6 +45,14 @@ const getSettersAndGetters = (
     }
 
+    if (isDecoratedAsExportArgs(setGet)) {
+      exportText = parseExportArgs(setGet, props)
```
I am wondering if we should streamline these 3 functions (2+one inline reponse) into one with arguments.

adamuso replied:
I think we should because the same code repeats in two places (in `parse_property_declaration.ts` and `parse_class_declaration.ts`)

ksjogo replied:
Sounds good 👍

johnfn commented:
Love the idea - I think it would be perfect with @ksjogo's suggestion about optional arguments for @exports().

adamuso commented:
Okay, so I've added few commits which address your comments about exports arguments. 

List of changes:
- `@export_args` and `@exports` are merged together, so old plain `@exports` works but also optionally it can take additional arguments
- `@exports` automatically infers and adds array element type with support of nested arrays and `any[]` or `unknown[]` arrays for old behavior which generates `export(Array) var ...`, now arrays like `float[]` generate `export(Array, float) var ...`
- `@exports` arguments can be strings or numbers which will be passed as they are to the gdscript `export`
```ts
@exports("abc", "def")
prop: string;
// generates
export(String, "abc", "def") var prop: String
``` 
- `@exports` arguments can also use special enum called `ExportHint` for adding gdscript specific exports (like `DIR`, `GLOBAL`, etc.)
```ts
@exports(ExportHint.EXP, 100, 1000, 20)
prop: float;
// generates
export(float, EXP, 100, 1000, 20) var prop: float
```
- we have already `PropertyHint` generated automatically but I did not want to use it because it is not compatible with gdscript exports (there are similarities but using it would need additional parsing/logic) that is why I've added `ExportHint` enum.

Please review the code when you will have some free time. :)

ksjogo commented:
Looks great! Will have a deeper look end of the week work for the year is finished.

johnfn commented:
This looks fantastic - you really went above and beyond here! I'll be able to review this more thoroughly either today or tomorrow.

johnfn commented:
Yeah, awesome stuff. As a followup, if you want to add this functionality to the README that could be really useful :)

adamuso commented:
I will update it in my free time 🙂
